.circleci/config.yml

```diff
@@ -3,67 +3,174 @@
-      - image: circleci/python:3.7.2
-
-      # Specify service dependencies here if necessary
-      # CircleCI maintains a library of pre-built images
-      # documented at https://circleci.com/docs/2.0/circleci-images/
-      # - image: circleci/postgres:9.4
+      - image: circleci/python:<< parameters.tag >>
       - image: circleci/buildpack-deps:stretch
-
     working_directory: ~/repo
 
-    steps:
-      - checkout
+commands:
 
-      - run:
-          name: setup docker prereqs
-          command: sudo apt-get update && sudo apt-get install -y --no-install-recommends
-            libudev-dev libavformat-dev libavcodec-dev libavdevice-dev libavutil-dev
-            libswscale-dev libswresample-dev libavfilter-dev
+  docker-prereqs:
+    description: Set up docker prerequisite requirement
+    steps:
+      - run: sudo apt-get update && sudo apt-get install -y --no-install-recommends
+          libudev-dev libavformat-dev libavcodec-dev libavdevice-dev libavutil-dev
+          libswscale-dev libswresample-dev libavfilter-dev
 
-      # Download and cache dependencies, we don't use fallback cache
+  install-requirments:
+    description: Set up venv and install requirements python packages with cache support
+    parameters:
+      python:
+        type: string
+        default: latest
+      all:
+        description: pip install -r requirements_all.txt
+        type: boolean
+        default: false
+      test:
+        description: pip install -r requirements_test.txt
+        type: boolean
+        default: false
+      test_all:
+        description: pip install -r requirements_test_all.txt
+        type: boolean
+        default: false
+    steps:
       - restore_cache:
           keys:
-            - v1-dependencies-{{ checksum "requirements_all.txt" }}-{{ checksum "requirements_test.txt" }}
-
+            - v1-<< parameters.python >>-<<# parameters.all >>{{ checksum "requirements_all.txt" }}<</ parameters.all>>-<<# parameters.test >>{{ checksum "requirements_test.txt" }}<</ parameters.test>>-<<# parameters.test_all >>{{ checksum "requirements_test_all.txt" }}<</ parameters.test_all>>
       - run:
           name: install dependencies
           command: |
             python3 -m venv venv
             . venv/bin/activate
-            pip install -q --progress-bar off -r requirements_all.txt -r requirements_test.txt -c homeassistant/package_constraints.txt
-
+            pip install -U pip
+            <<# parameters.all >>pip install -q --progress-bar off -r requirements_all.txt -c homeassistant/package_constraints.txt<</ parameters.all>>
+            <<# parameters.test >>pip install -q --progress-bar off -r requirements_test.txt -c homeassistant/package_constraints.txt<</ parameters.test>>
+            <<# parameters.test_all >>pip install -q --progress-bar off -r requirements_test_all.txt -c homeassistant/package_constraints.txt<</ parameters.test_all>>
```

Comment:
drop this.

Reply:
This is conditional line controlled by boolean parameters. test_all is required for test, test+all is required for pylint. They are different. So all three parameters are need for here.